hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/RegionRemoteProcedureBase.java

```diff
@@ -296,12 +296,25 @@ private void setFuture(CompletableFuture<Void> f) {
   }
 
   @Override
-  protected Procedure<MasterProcedureEnv>[] execute(MasterProcedureEnv env)
-    throws ProcedureYieldException, ProcedureSuspendedException, InterruptedException {
+  protected void beforeExec(MasterProcedureEnv env) throws ProcedureSuspendedException {
     RegionStateNode regionNode = getRegionNode(env);
     if (!regionNode.isLockedBy(this)) {
       regionNode.lock(this, () -> ProcedureFutureUtil.wakeUp(this, env));
     }
+  }
+
+  @Override
+  protected void afterExec(MasterProcedureEnv env) {
+    if (future == null) {
```

Comment:
I guess we did so before, but it sure looks odd to only conditionally release the lock.

Reply:
This is an improvement on 3.x, where we will throw ProcedureSuspendedException when updating meta, to avoid consuming a PEWorker while doing external I/O. If the future is not null, it means we have a out going meta updating operation which is not finished yet, so we should not release the region node lock, to keep the old locking way.

Please see HBASE-28196.